File: ios/Classes/Transfer/TransactionParamsMapper.m
#import "TransactionParamsMapper.h"

@implementation TransactionParamsMapper

+ (P24TransactionParams *) buildTransaction: (NSDictionary *) transactionParams {
    P24TransactionParams *p24Payment = [P24TransactionParams new];
    
     int merchantId = [[transactionParams valueForKey:@"merchantId"] intValue];
     NSString* crc = [transactionParams objectForKey:@"crc"];
     NSString* sessionId = [transactionParams objectForKey:@"sessionId"];
     int amount = [[transactionParams valueForKey:@"amount"] intValue];
     NSString* currency = [transactionParams objectForKey:@"currency"];
     NSString* description = [transactionParams objectForKey:@"description"];
     NSString* email = [transactionParams objectForKey:@"email"];
     NSString* country = [transactionParams objectForKey:@"country"];
     NSString* client = [transactionParams objectForKey:@"client"];
     NSString* address = [transactionParams objectForKey:@"address"];
     NSString* zip = [transactionParams objectForKey:@"zip"];
     NSString* city = [transactionParams objectForKey:@"city"];
     NSString* phone = [transactionParams objectForKey:@"phone"];
     NSString* language = [transactionParams objectForKey:@"language"];
     int method = [[transactionParams valueForKey:@"method"] intValue];
     NSString* urlStatus = [transactionParams objectForKey:@"urlStatus"];
     int timeLimit = [[transactionParams valueForKey:@"timeLimit"] intValue];
     int channel = [[transactionParams valueForKey:@"channel"] intValue];
     int shipping = [[transactionParams valueForKey:@"shipping"] intValue];
     NSString* transferLabel = [transactionParams objectForKey:@"transferLabel"];
     
    NSArray* itemsList = [transactionParams objectForKey:@"passageCart"];
    
    if(merchantId) { p24Payment.merchantId = merchantId; }
    if(crc) { p24Payment.crc = crc; }
    if(sessionId) { p24Payment.sessionId = sessionId; }
    if(amount) { p24Payment.amount = amount; }
    if(currency) { p24Payment.currency = currency; }
    if(description) { p24Payment.desc = description; }
    if(email) { p24Payment.email = email; }
    if(country) { p24Payment.country = country; }
    if(client) { p24Payment.client = client; }
    if(address) { p24Payment.address = address; }
    if(zip) { p24Payment.zip = zip; }
    if(city) { p24Payment.city = city; }
    if(phone) { p24Payment.phone = phone; }
    if(language) { p24Payment.language = language; }
    if(method) { p24Payment.method = method; }
    if(urlStatus) { p24Payment.urlStatus = urlStatus; }
    if(timeLimit) { p24Payment.timeLimit = timeLimit; }
    if(channel) { p24Payment.channel = channel; }
    if(shipping) { p24Payment.shipping = shipping; }
    if(transferLabel) { p24Payment.transferLabel = transferLabel; }

    if (itemsList || itemsList.count){
        P24PassageCart * cart = [self buildPassageCart: itemsList];
        p24Payment.passageCart = cart;
        p24Payment.amount = [cart getSummaryPrice];
    }
    
    return p24Payment;
}

+ (P24PassageCart *) buildPassageCart: (NSArray *) itemsList {
    P24PassageCart* passageCart = [P24PassageCart new];
      int counter = 1;
      for (id item in itemsList) {
          P24PassageItem* passageItem = [self buildPassageItem: item];
          [passageCart addItem: passageItem];
          counter++;
      }
    return passageCart;
}

+ (P24PassageItem *) buildPassageItem: (NSDictionary *) dictionary {
    P24PassageItem* passageItem = [[P24PassageItem alloc] initWithName: @"item"];

    NSString* name = [dictionary objectForKey:@"name"];
    NSString* description = [dictionary objectForKey:@"description"];
    int quantity = [[dictionary valueForKey:@"quantity"] intValue];
    int price = [[dictionary valueForKey:@"price"] intValue];
    int number = [[dictionary valueForKey:@"number"] intValue];
    int targetAmount = [[dictionary valueForKey:@"targetAmount"] intValue];
    int targetPosId = [[dictionary valueForKey:@"targetPosId"] intValue];
    
    passageItem.name = name;
    passageItem.desc = description;
    passageItem.quantity = quantity;
    passageItem.price = price;
    passageItem.number = number;
    passageItem.targetAmount = targetAmount;
    passageItem.targetPosId = targetPosId;
    
    return passageItem;
}
@end

File: ios/Classes/Transfer/TransactionParamsMapper.swift
import P24

internal struct TransactionParamsMapper {

    internal static func buildTransaction(_ transactionParams: [String: Any]) -> P24TransactionParams {

        let p24Payment = P24TransactionParams()
        
        p24Payment.merchantId = transactionParams["merchantId"] as! Int32
        p24Payment.crc = transactionParams["crc"] as? String
        p24Payment.sessionId = transactionParams["sessionId"] as? String
        p24Payment.amount = transactionParams["amount"] as! Int32
        p24Payment.currency = transactionParams["currency"] as? String
        p24Payment.desc = transactionParams["description"] as? String
        p24Payment.email = transactionParams["email"] as? String
        
        p24Payment.country = transactionParams["country"] as? String
        p24Payment.client = transactionParams["client"] as? String
        p24Payment.address = transactionParams["address"] as? String
        p24Payment.zip = transactionParams["zip"] as? String
        p24Payment.city = transactionParams["city"] as? String
        p24Payment.phone = transactionParams["phone"] as? String
        p24Payment.language = transactionParams["language"] as? String
        p24Payment.method = transactionParams["method"] as! Int32
        p24Payment.language = transactionParams["language"] as? String
        p24Payment.urlStatus = transactionParams["urlStatus"] as? String
        p24Payment.transferLabel = transactionParams["transferLabel"] as? String

        if let timeLimit = transactionParams["timeLimit"] as? Int32 { p24Payment.timeLimit = timeLimit }
        if let channel = transactionParams["channel"] as? Int32 { p24Payment.channel = channel }
        if let shipping = transactionParams["shipping"] as? Int32 { p24Payment.shipping = shipping}
        
        if let itemsList = transactionParams["passageCart"] as? NSArray, itemsList.count > 0 {
            let cart = buildPassageCart(itemsList)
            p24Payment.passageCart = cart
            p24Payment.amount = cart.getSummaryPrice()
        }

        return p24Payment
    }

    private static func buildPassageCart(_ itemsList: NSArray) -> P24PassageCart {
        let passageCart = P24PassageCart()
        itemsList.forEach { item in
            let passageItem = buildPassageItem(item as! NSDictionary)
            passageCart.addItem(passageItem)
        }
        return passageCart
    }

    private static func buildPassageItem(_ dictionary: NSDictionary) -> P24PassageItem {
        let passageItem = P24PassageItem()
        passageItem.name = dictionary["name"] as? String
        passageItem.desc = dictionary["description"] as? String
        passageItem.quantity = dictionary["quantity"] as! Int32
        passageItem.price = dictionary["price"] as! Int32
        passageItem.number = dictionary["number"] as! Int32
        passageItem.targetAmount = dictionary["targetAmount"] as! Int32
        passageItem.targetPosId = dictionary["targetPosId"] as! Int32
        return passageItem
    }
}

